### PR TITLE
Fix window focus detection on Linux (rebase from #3843)

### PR DESCRIPTION
**Pull request by aldelaro5**
This is a rebase I did from pr #3843 with the permission of the author.

From my understanding, for some reasons, if you are in some environement (I know so far my arch linux, linux mint 17.x + I think 16 and ubuntu 14.04 with unity), the FindFocus function from wx just doesn't work well, it seems to tell dolphin has the focus even though it wouldn't be the case.  This would cause 2 main issues:

* Pause on focus lost simply never works
* Hotkeys and input would be processed by dolphin even when it doesn't have the focus and background input disabled

Compared to the original PR, what I did is I addressed the last unresolved review, but most of the original pr is there.  Basically, the idea is to set a flag when the main frame gets activated and the helper method would return just that.

I also addressed the first and third points of @EmptyChaos 's [comment](https://github.com/dolphin-emu/dolphin/pull/3843#issuecomment-220885493) here which is the last comment left unaddressed before the pr was no longer maintained. ~~The only point left to be addresses before considering for merge is about UiHasFocus.~~ This has been addressed so it is ready for review and merge.

Which brings me to some questions for @EmptyChaos 

~~Can you confirm your comment still applies?  This pr was pretty old, I had to even resolve conflicts with formatting because of clang format which means there may be changes I am not aware of that would changes things.  Also, in the case it still needs to be addressed, can you please explain me how I would implement UiHasFocus using a wx activate event?  I am not sure how it works other than those who needs UiHasFocus needs to handle the event.  I found the comment slightly vague.~~

EDIT: I actually kinda figured out how it works (and if I got it right, it's much mroe simple than I thought), but you should probably review it to be sure.

Tested on a Windows 10 VM to be sure, didn't found any apparent regression.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4394)

### CODE REVIEW

File: Source/Core/DolphinWX/Frame.h [hunk 1]
```diff
@@ -158,6 +158,7 @@ class CFrame : public CRenderFrame
   bool m_bNoDocking = false;
   bool m_bGameLoading = false;
   bool m_bClosing = false;
+  bool m_bHasFocus;
```
Comment (lioncash) [inline Source/Core/DolphinWX/Frame.h]: This should have a default value like the others.

Comment (aldelaro5) [inline Source/Core/DolphinWX/Frame.h]: Done.

**Comment by lioncash** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 798 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KV98WrvY6wJGAJW0jrk:-KV98Wrwct0SrCWYpSA_:b5mur4q) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ce2c5d44acfb8f211e3ac9f685d4b66de329b7e3/Source/Core/DolphinWX/Frame.cpp#L798)):*
> ```cpp
> }
> 
> void CFrame::SetUiFocusStatus(bool hasFocus)
> ```

This should be done by posting an event to the frame, not by the main application instance directly calling into it.

---

*[Source/Core/DolphinWX/Frame.h, line 161 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KV7i2twz9TzC6VFefGW-r1-161:-KV985UayIBDtd6QQNKv:b-ecizlm) ([raw file](https://github.com/dolphin-emu/dolphin/blob/dbdc87fa113c9ca0b3aba527c6de0976d58d74a5/Source/Core/DolphinWX/Frame.h#L161)):*
<details><summary><i>Previously, aldelaro5 (aldelaro5) wrote…</i></summary>
> Done.
</details>
You should probably by indicating what the difference is between m_bHasFocus and m_bUIHasFocus.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by BhaaLseN** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.h, line 161 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KV7i2twz9TzC6VFefGW-r1-161:-KVA9O2jLkupD-8zPCdj:b-5lhkem) ([raw file](https://github.com/dolphin-emu/dolphin/blob/dbdc87fa113c9ca0b3aba527c6de0976d58d74a5/Source/Core/DolphinWX/Frame.h#L161)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You should probably by indicating what the difference is between m_bHasFocus and m_bUIHasFocus.
</details>
Might be worthwhile renaming it to `m_bRendererHasFocus` to link them more closely to their usage within `CFrame::RendererHasFocus` and `CFrame::UIHasFocus`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by aldelaro5** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 798 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KV98WrvY6wJGAJW0jrk:-KVAMuBbDeFYKuWweSA9:bhof1m6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ce2c5d44acfb8f211e3ac9f685d4b66de329b7e3/Source/Core/DolphinWX/Frame.cpp#L798)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should be done by posting an event to the frame, not by the main application instance directly calling into it.
</details>
Will do later.

---

*[Source/Core/DolphinWX/Frame.h, line 161 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KV7i2twz9TzC6VFefGW-r1-161:-KVAN1Mqg_0C9tijbg9V:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/dbdc87fa113c9ca0b3aba527c6de0976d58d74a5/Source/Core/DolphinWX/Frame.h#L161)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> Might be worthwhile renaming it to `m_bRendererHasFocus` to link them more closely to their usage within `CFrame::RendererHasFocus` and `CFrame::UIHasFocus`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by aldelaro5** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 798 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KV98WrvY6wJGAJW0jrk:-KVAgYAGuts0e2EuuLU4:b-cy5d0j) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ce2c5d44acfb8f211e3ac9f685d4b66de329b7e3/Source/Core/DolphinWX/Frame.cpp#L798)):*
<details><summary><i>Previously, aldelaro5 (aldelaro5) wrote…</i></summary>
> Will do later.
</details>
Done.

Trying to get pause on focus lost working, but it is kinda hard because FindFocus just doesn't work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by aldelaro5** [comment]
Ok, after spending some hours I finalyl figured out all this confusion.

Basically, this pr is now ready for review and merge.

There was some issues I found with pause on focus lost under my linux.  Namely, it would be too strict like right clicking on the code view, moving the window and the little menu that appears on the right of the toolbar when the window isn;t wide enough to contain all of them (it doesn't seem to appear on Windows).  For the toolbar expansion menu, it actually would cause a focus lost which would cause a required UpdateGui to happen which screws up that menu so it jsut doesn't work. 

After spending some time, I found out that these issues aren't related to this pr at all, they were there since then on master.  Because of this, these issues (likely cause by wx's weird management of focus) aren't in the scope of this pr to fix it, I tried still, but it really seems that I can't.  Anyway this is a different issue.

The rest works as intended though so this pr is now up for review, it fixes very annoying bugs that finally won't annoy me as much :)

**Comment by EmptyChaos** [comment]
The problem with menus, (I assume you're using Render To Main Window) is that X11 doesn't have a concept of "menu window", they're real toplevel desktop windows which steal focus. There doesn't seem to be any way to fix this, I've been messing around with `gtk_window_list_toplevels`, `gtk_window_is_active` and `gtk_window_has_toplevel_focus` and none of them seem to work while a popup window is open.

---

Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 516 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBRKTcy-LSAdwGUnMb:-KVBRKTcy-LSAdwGUnMc:b-4e30pl) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Frame.cpp#L516)):*
> ```cpp
> 
>       if (SConfig::GetInstance().m_PauseOnFocusLost && Core::GetState() == Core::CORE_PAUSE)
>         Core::SetState(Core::CORE_RUN);
> ```

Use `DoPause()` instead that way you don't need to call `UpdateGUI` explicitly.

---

*[Source/Core/DolphinWX/Frame.cpp, line 524 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBRUsbCK-CGmYVpJFA:-KVBRUscIXpTG8rvF-k2:b39biz3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Frame.cpp#L524)):*
> ```cpp
>     {
>       if (SConfig::GetInstance().m_PauseOnFocusLost && Core::GetState() == Core::CORE_RUN)
>         Core::SetState(Core::CORE_PAUSE);
> ```

`DoPause()`

---

*[Source/Core/DolphinWX/Frame.cpp, line 795 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBSIuVYRE2c6oe_vER:-KVBSIuVYRE2c6oe_vES:bkj3hox) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Frame.cpp#L795)):*
> ```cpp
>     return true;
> #else
>   return m_bRendererHasFocus;
> ```

You don't need separate cases for Windows and everything else, the `#ifdef` can go. `WM_ACTIVATE` is a Win32 UI concept that wxWidgets makes available on other platforms.

---

*[Source/Core/DolphinWX/FrameTools.cpp, line 929 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBR3a-e0S2A8-KktqR:-KVBR3a-e0S2A8-KktqS:b-co3ji3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/FrameTools.cpp#L929)):*
> ```cpp
>   }
>   m_RenderParent = nullptr;
>   m_bRendererHasFocus = false;
> ```

You need to set `m_RenderFrame` to `nullptr` as well. Master just leaves the value set to a dangling pointer.

---

*[Source/Core/DolphinWX/Main.cpp, line 260 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBREzHj46-ZiZpZiB8:-KVBREzI_sOhPkjIYABV:b-vxbip8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Main.cpp#L260)):*
> ```cpp
> }
> 
> void DolphinApp::OnActivate(wxActivateEvent& event)
> ```

This looks mostly right but the way you're handling `wxEVT_ACTIVATE_APP` is overly complex.
```cpp
class DolphinApp {
public:
   // ...
   bool IsActiveThreadsafe() const { return m_is_active; }
   // ...
private:
   // ...
   bool m_is_active = true;
};

void DolphinApp::OnActivate(wxActivateEvent& ev)
{
   m_is_active = ev.GetActive();
}

bool Host_UIHasFocus()
{
   return wxGetApp().IsActiveThreadsafe();
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-:-KVBqFIQjoqILtiO64fM:b-syzchl)*

**Comment by aldelaro5** [comment]
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 516 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBRKTcy-LSAdwGUnMb:-KVCULvQabF1NWqALMQ6:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Frame.cpp#L516)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> Use `DoPause()` instead that way you don't need to call `UpdateGUI` explicitly.
</details>
Done.

---

*[Source/Core/DolphinWX/Frame.cpp, line 524 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBRUsbCK-CGmYVpJFA:-KVCUM_lpfbGJfT9GyFq:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Frame.cpp#L524)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> `DoPause()`
</details>
Done.

---

*[Source/Core/DolphinWX/Frame.cpp, line 795 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBSIuVYRE2c6oe_vER:-KVCUOsmI7By-kEJcov4:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Frame.cpp#L795)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> You don't need separate cases for Windows and everything else, the `#ifdef` can go. `WM_ACTIVATE` is a Win32 UI concept that wxWidgets makes available on other platforms.
</details>
Done.

---

*[Source/Core/DolphinWX/FrameTools.cpp, line 929 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBR3a-e0S2A8-KktqR:-KVCUS-1KjRzRPdclbdj:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/FrameTools.cpp#L929)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> You need to set `m_RenderFrame` to `nullptr` as well. Master just leaves the value set to a dangling pointer.
</details>
Done.

---

*[Source/Core/DolphinWX/Main.cpp, line 260 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVBREzHj46-ZiZpZiB8:-KVCUZ3_TsU2DFVn-jmY:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9b3a5f6f4d76de280036067afdc7f8405a842ba5/Source/Core/DolphinWX/Main.cpp#L260)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> This looks mostly right but the way you're handling `wxEVT_ACTIVATE_APP` is overly complex.
> ```cpp
> class DolphinApp {
> public:
>    // ...
>    bool IsActiveThreadsafe() const { return m_is_active; }
>    // ...
> private:
>    // ...
>    bool m_is_active = true;
> };
> 
> void DolphinApp::OnActivate(wxActivateEvent& ev)
> {
>    m_is_active = ev.GetActive();
> }
> 
> bool Host_UIHasFocus()
> {
>    return wxGetApp().IsActiveThreadsafe();
> }
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by EmptyChaos** [comment]
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 682 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVCgQP2WVaTlAG5VTfi:-KVCgQP2WVaTlAG5VTfj:bddffvb) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c166144ff51b29ff6a9e11ceedb7faaaf49c6c67/Source/Core/DolphinWX/Frame.cpp#L682)):*
> ```cpp
>     break;
> 
>   case IDM_UI_FOCUS_CHANGE:
> ```

This isn't needed now.

---

*[Source/Core/DolphinWX/Frame.cpp, line 791 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVCfXWvDJ5FC6JTPZ0v:-KVCfXWvDJ5FC6JTPZ0w:baz7c33) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c166144ff51b29ff6a9e11ceedb7faaaf49c6c67/Source/Core/DolphinWX/Frame.cpp#L791)):*
> ```cpp
> // Returns true any time any one of our UI windows
> // has the focus, including any dialogs or other windows.
> bool CFrame::UIHasFocus()
> ```

Since you implemented my suggestion of moving this into `DolphinApp`, nothing calls this function any more. You can delete it (and the `m_bUIHasFocus` member variable)

---

*[Source/Core/DolphinWX/Globals.h, line 303 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVCfyM9wuZgnFHrUQ72:-KVCfyM9wuZgnFHrUQ73:bddffvb) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c166144ff51b29ff6a9e11ceedb7faaaf49c6c67/Source/Core/DolphinWX/Globals.h#L303)):*
> ```objc
>   IDM_PANIC,
>   IDM_KEYSTATE,
>   IDM_UI_FOCUS_CHANGE,
> ```

This isn't needed now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by aldelaro5** [comment]
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/DolphinWX/Frame.cpp, line 682 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVCgQP2WVaTlAG5VTfi:-KVCkaEwwzpymagrsmsR:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c166144ff51b29ff6a9e11ceedb7faaaf49c6c67/Source/Core/DolphinWX/Frame.cpp#L682)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> This isn't needed now.
</details>
Done.

---

*[Source/Core/DolphinWX/Frame.cpp, line 791 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVCfXWvDJ5FC6JTPZ0v:-KVCkbWxf65I2cDxWO-w:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c166144ff51b29ff6a9e11ceedb7faaaf49c6c67/Source/Core/DolphinWX/Frame.cpp#L791)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> Since you implemented my suggestion of moving this into `DolphinApp`, nothing calls this function any more. You can delete it (and the `m_bUIHasFocus` member variable)
</details>
Done.

---

*[Source/Core/DolphinWX/Globals.h, line 303 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-KVCfyM9wuZgnFHrUQ72:-KVCkdsX8FhCrgMCgBZw:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c166144ff51b29ff6a9e11ceedb7faaaf49c6c67/Source/Core/DolphinWX/Globals.h#L303)):*
<details><summary><i>Previously, EmptyChaos wrote…</i></summary>
> This isn't needed now.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394)*

**Comment by EmptyChaos** [comment]
:lgtm:

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4394#-:-KVCl5C6hZ_nmzrzn7-U:bnfp4nl)*

**Comment by Ebola16** [comment]
This PR is likely the cause of [Disabling background input and minimizing game window after losing focus still allows background input](https://bugs.dolphin-emu.org/issues/10651)

**Pull request merged**